pkg/sql/sem/tree/casts.go

```diff
@@ -0,0 +1,1012 @@
+	{from: types.DateFamily, to: types.TimestampTZFamily, volatility: VolatilityStable},
+	{from: types.TimestampFamily, to: types.TimestampTZFamily, volatility: VolatilityStable},
+	{from: types.TimestampTZFamily, to: types.TimestampTZFamily, volatility: VolatilityImmutable},
+	{from: types.IntFamily, to: types.TimestampTZFamily, volatility: VolatilityImmutable},
+
+	// Casts to IntervalFamily.
+	{from: types.UnknownFamily, to: types.IntervalFamily, volatility: VolatilityImmutable},
+	{from: types.StringFamily, to: types.IntervalFamily, volatility: VolatilityImmutable},
+	{from: types.CollatedStringFamily, to: types.IntervalFamily, volatility: VolatilityImmutable},
+	{from: types.IntFamily, to: types.IntervalFamily, volatility: VolatilityImmutable},
+	{from: types.TimeFamily, to: types.IntervalFamily, volatility: VolatilityImmutable},
+	{from: types.IntervalFamily, to: types.IntervalFamily, volatility: VolatilityImmutable},
+	{from: types.FloatFamily, to: types.IntervalFamily, volatility: VolatilityImmutable},
+	{from: types.DecimalFamily, to: types.IntervalFamily, volatility: VolatilityImmutable},
+
+	// Casts to OidFamily.
+	{from: types.UnknownFamily, to: types.OidFamily, volatility: VolatilityImmutable},
+	{from: types.StringFamily, to: types.OidFamily, volatility: VolatilityStable},
+	{from: types.CollatedStringFamily, to: types.OidFamily, volatility: VolatilityStable},
+	{from: types.IntFamily, to: types.OidFamily, volatility: VolatilityStable},
+	{from: types.OidFamily, to: types.OidFamily, volatility: VolatilityStable},
+
+	// Casts to UuidFamily.
+	{from: types.UnknownFamily, to: types.UuidFamily, volatility: VolatilityImmutable},
+	{from: types.StringFamily, to: types.UuidFamily, volatility: VolatilityImmutable},
+	{from: types.CollatedStringFamily, to: types.UuidFamily, volatility: VolatilityImmutable},
+	{from: types.BytesFamily, to: types.UuidFamily, volatility: VolatilityImmutable},
+	{from: types.UuidFamily, to: types.UuidFamily, volatility: VolatilityImmutable},
+
+	// Casts to INetFamily.
+	{from: types.UnknownFamily, to: types.INetFamily, volatility: VolatilityImmutable},
+	{from: types.StringFamily, to: types.INetFamily, volatility: VolatilityImmutable},
+	{from: types.CollatedStringFamily, to: types.INetFamily, volatility: VolatilityImmutable},
+	{from: types.INetFamily, to: types.INetFamily, volatility: VolatilityImmutable},
+
+	// Casts to ArrayFamily.
+	{from: types.UnknownFamily, to: types.ArrayFamily, volatility: VolatilityImmutable},
+	{from: types.StringFamily, to: types.ArrayFamily, volatility: VolatilityStable},
+
+	// Casts to JsonFamily.
+	{from: types.UnknownFamily, to: types.JsonFamily, volatility: VolatilityImmutable},
+	{from: types.StringFamily, to: types.JsonFamily, volatility: VolatilityImmutable},
+	{from: types.JsonFamily, to: types.JsonFamily, volatility: VolatilityImmutable},
+
+	// Casts to EnumFamily.
+	{from: types.UnknownFamily, to: types.EnumFamily, volatility: VolatilityImmutable},
```

Comment:
This is missing a bytes -> enum cast

Reply:
Will fix, I resolved conflicts with that change but I guess I added the opposite cast. Does seem like this is missing tests (I see you added a logictest in a subsequent PR but other casts have tests inside `sem/tree`).

Reply:
Yeah, i would like a test here, but it is a little tricky simulate the user defined type without the machinery creating it in the background. I'll spend some time making this easier to do.